docs/design/node_taint_eviction.md

```diff
@@ -42,51 +48,42 @@ automaticReplacementOptions:
+
+**Requirement 3.** The feature lists the taint status for process groups. This allows users to have a holistic view of which Pods in a FDB cluster are running on tainted Nodes and being replaced.
+
+### Implementation
+We need to make the following four type of changes:
+
+**CRD change.** We need to extend the `automaticReplacementOptions` with the new fields `taints` and `taintReplacementTimeSeconds`
+
+**Detect Node taint and change Pod condition.**  We can piggy-back the existing `validateProcessGroup()` function in `update_status` reconciler to detect if a Pod is running on a tainted Node and mark the Pod's taint-related condition accordingly.
+
+A Pod's Node can be listed by
 ```
+node = &corev1.Node{}
+err := r.Get(ctx, client.ObjectKey{Name: pod.Spec.NodeName}, node)
+```
+
+**Status of Pod taint-related condition.** We will add two  [ProcessGroupConditionType](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/api/v1beta2/foundationdbcluster_types.go#L673): `NodeTaintDetected`, which represents a Pod's Node is tainted but not long enough for operator to replace it, and `NodeTaintReplacing`, which represents a Pod whose Node has been tainted for longer-than-configured time and the operator should replace it.
+
+**Replace Pods with `NodeTaintReplacing` conditioin.** We can piggy-back existing replacement logic to replace Pods with `NodeTaintReplacing` condition. We can do this by adding `NodeTaintReplacing` to [`conditionsThatNeedReplacement` list](https://github.com/FoundationDB/fdb-kubernetes-operator/blob/d38d6bc7abf7764215976b75d581c933280389f8/api/v1beta2/foundationdbcluster_types.go#L65-L67)
+
+
+## Limitations
+The operator will only detect tainted Nodes when its reconciliation is triggered periodically. This may increase the delay of detecting and replacing Pods on tainted Nodes.
+
+The limitation can be resolved by letting the operator watch node events. The controller-runtime already supports to let the operator watch resources that are not managed by itself, see [Watching Externally Managed Resources](https://book.kubebuilder.io/reference/watching-resources/externally-managed.html).
 
-The limitation of this approach would be that we only detect Pods running on tainted nodes when a reconciliation is happening (those are not coupled to node events).
-To work around this we could extend the operator and let the operator watch node events, that could be potentially a high volume of events, so we have to make sure we get the right signals.
-The controller-runtime already supports to let the operator watch resources that are not managed by itself, see [Watching Externally Managed Resources](https://book.kubebuilder.io/reference/watching-resources/externally-managed.html).
+Because a k8s cluster may have a high volume of node events, we must make sure we get the right signals and not overload the operator by excessive events.
```

Comment:
This is a limitation? I would rather see this as a requirement. Is this statement part of the sentence above?

Reply:
yes

Reply:
removed this sentecne